contracts/diaspore/cosigner/pawn/Poach.sol

```diff
@@ -0,0 +1,149 @@
+        Pair storage pair = poaches[_id];
+        return (pair.token, pair.balance);
+    }
+
+    /**
+        @notice Create a pair and push into the poaches array
+
+        @param _token Token address (ERC20)
+        @param _amount Token amount
+
+        @return _id Index of pair in the poaches array
+    */
+    function create(
+        Token _token,
+        uint256 _amount
+    ) external payable returns (uint256 _id) {
+        _deposit(_token, _amount);
+
+        _id = poaches.push(Pair(_token, _amount)) - 1;
+
+        _generate(_id, msg.sender);
+
+        emit Created(_id, msg.sender, _token, _amount);
+    }
+
+    /**
+        @notice Deposit an amount of token in a pair
+
+        @dev If the currency its ether and the destiny its a contract, execute the payable deposit()
+
+        @param _id Index of pair in poaches array
+        @param _amount Token amount
+
+        @return true If the operation was executed
+    */
+    function deposit(
+        uint256 _id,
+        uint256 _amount
+    ) external payable onlyAuthorized(_id) returns (bool) {
+        Pair storage pair = poaches[_id];
+        _deposit(pair.token, _amount);
+
+        pair.balance += _amount;
+
+        emit Deposit(_id, msg.sender, _amount);
+
+        return true;
+    }
+
+    /**
+        @notice Withdraw all funds of a pair
+
+        @param _id Index of pair in poaches array
+        @param _to The beneficiary of returned funds
+
+        @return true If the operation was executed
+    */
+    function withdraw(
+        uint256 _id,
+        address payable _to
+    ) external onlyAuthorized(_id) returns (bool) {
+        require(_to != address(0), "_to should not be 0x0");
+        Pair storage pair = poaches[_id];
+
+        uint256 balance = pair.balance;
+
+        if (pair.token != Token(ETH))
+            require(pair.token.transfer(_to, balance), "Error transfer tokens");
+        else
+            _to.transfer(balance);
+
+        delete (pair.balance);
+
+        emit Withdraw(_id, msg.sender, _to, balance);
+
+        return true;
+    }
+
+    /**
+        @notice Withdraw a partial amount of funds of a pair
+
+        @param _id Index of pair in poaches array
+        @param _to The beneficiary of returned funds
+        @param _amount The amount of returned funds
+
+        @return true If the operation was executed
+    */
+    function withdrawPartial(
+        uint256 _id,
+        address payable _to,
+        uint256 _amount
+    ) external onlyAuthorized(_id) returns (bool) {
+        require(_to != address(0), "_to should not be 0x0");
+        Pair storage pair = poaches[_id];
+
+        require(pair.balance >= _amount, "The balance of poach its to low");
+
+        if (pair.token != Token(ETH))
+            require(pair.token.transfer(_to, _amount), "Error transfer tokens");
+        else
+            _to.transfer(_amount);
+
+        pair.balance -= _amount;
```

Comment:
this sentence should be before the transfer.

Reply:
and use SafeMath

Reply:
`this sentence should be before the transfer` Good approach!
I dont use SafeMath becouse i check the underflow before the substract

Reply:
=)